src/main/telemetry/msp_shared.c

```diff
@@ -0,0 +1,231 @@
+#include <stdbool.h>
+#include <stdint.h>
+#include <stdlib.h>
+#include <math.h>
+#include <string.h>
+
+#include "platform.h"
+
+#ifdef TELEMETRY
+
+#include "build/build_config.h"
+
+#include "common/utils.h"
+
+#include "fc/fc_msp.h"
+
+#include "msp/msp.h"
+
+#include "rx/crsf.h"
+#include "rx/msp.h"
+
+#include "telemetry/msp_shared.h"
+#include "telemetry/smartport.h"
+
+#define TELEMETRY_MSP_VERSION    1
+#define TELEMETRY_MSP_VER_SHIFT  5
+#define TELEMETRY_MSP_VER_MASK   (0x7 << TELEMETRY_MSP_VER_SHIFT)
+#define TELEMETRY_MSP_ERROR_FLAG (1 << 5)
+#define TELEMETRY_MSP_START_FLAG (1 << 4)
+#define TELEMETRY_MSP_SEQ_MASK   0x0F
+#define TELEMETRY_MSP_RES_ERROR (-10)
+
+enum {
+    TELEMETRY_MSP_VER_MISMATCH=0,
+    TELEMETRY_MSP_CRC_ERROR=1,
+    TELEMETRY_MSP_ERROR=2
+};
+
+#define REQUEST_BUFFER_SIZE 64
+#define RESPONSE_BUFFER_SIZE 64
+
+STATIC_UNIT_TESTED uint8_t checksum = 0;
+STATIC_UNIT_TESTED mspPackage_t mspPackage;
+static mspRxBuffer_t mspRxBuffer;
+static mspTxBuffer_t mspTxBuffer;
+static mspPacket_t mspRxPacket;
+static mspPacket_t mspTxPacket;
+
+void initSharedMsp() {
+    mspPackage.requestBuffer = (uint8_t *)&mspRxBuffer;
+    mspPackage.requestPacket = &mspRxPacket;
+    mspPackage.requestPacket->buf.ptr = mspPackage.requestBuffer;
+    mspPackage.requestPacket->buf.end = mspPackage.requestBuffer;
+
+    mspPackage.responseBuffer = (uint8_t *)&mspTxBuffer;
+    mspPackage.responsePacket = &mspTxPacket;
+    mspPackage.responsePacket->buf.ptr = mspPackage.responseBuffer;
+    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+}
+
+static void processMspPacket()
+{
+    mspPackage.responsePacket->cmd = 0;
+    mspPackage.responsePacket->result = 0;
+    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+
+    mspPostProcessFnPtr mspPostProcessFn = NULL;
+    if (mspFcProcessCommand(mspPackage.requestPacket, mspPackage.responsePacket, &mspPostProcessFn) == MSP_RESULT_ERROR) {
+        sbufWriteU8(&mspPackage.responsePacket->buf, TELEMETRY_MSP_ERROR);
+    }
+    if (mspPostProcessFn) {
+        mspPostProcessFn(NULL);
+    }
+
+    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+}
+
+void sendMspErrorResponse(uint8_t error, int16_t cmd)
+{
+    mspPackage.responsePacket->cmd = cmd;
+    mspPackage.responsePacket->result = 0;
+    mspPackage.responsePacket->buf.end = mspPackage.responseBuffer;
+
+    sbufWriteU8(&mspPackage.responsePacket->buf, error);
+    mspPackage.responsePacket->result = TELEMETRY_MSP_RES_ERROR;
+    sbufSwitchToReader(&mspPackage.responsePacket->buf, mspPackage.responseBuffer);
+}
+
+bool handleMspFrame(uint8_t *frameStart, uint8_t *frameEnd)
+{
+    static uint8_t mspStarted = 0;
+    static uint8_t lastSeq = 0;
+
+    if (sbufBytesRemaining(&mspPackage.responsePacket->buf) > 0) {
+        return false;
```

Comment:
During hardware debugging, I found that the state is getting stuck at this position.  It appears that when the previous set of frames do not complete transport, the buffer is never fully read to the end point. The false return value tells smartport.c that the request was not handled and not to request a response.  The purpose of this check is to prevent answering requests when a current one is underway.  Perhaps a check for a pending response in smartport.c before requesting a new response would solve the issue.

Reply:
Good find!

The behaviour of the existing SmartPort implementation seems to be that a new request simply overrides the request that is currently processing, which seems to make more sense then dropping the later request (assuming that all MSP over SmartPort clients treat MSP requests as blocking, so that multiple MSP requests without responses will only happen for cases like program termination and restart).

Alternatively, queueing could be implemented for MSP requests over SmartPort, but it should not be necessary.

Reply:
I simply put a false check on smartPortMspReplyPending in smartport.c and the issue is resolved.. amend coming.